token-swap/program/src/processor.rs

```diff
@@ -7421,4 +7436,460 @@ mod tests {
+            owner_trade_fee_numerator: 1,
+            owner_trade_fee_denominator: 10,
+            owner_withdraw_fee_numerator: 1,
+            owner_withdraw_fee_denominator: 5,
+            host_fee_numerator: 7,
+            host_fee_denominator: 100,
+        };
+
+        let token_a_amount = 1000;
+        let token_b_amount = 2000;
+        let swap_curve = SwapCurve {
+            curve_type: CurveType::ConstantProduct,
+            calculator: Arc::new(ConstantProductCurve {}),
+        };
+
+        let withdrawer_key = Pubkey::new_unique();
+        let initial_a = token_a_amount / 10;
+        let initial_b = token_b_amount / 10;
+        let initial_pool = swap_curve.calculator.new_pool_supply() / 10;
+        let withdraw_amount = initial_pool / 4;
+        let minimum_token_a_amount = initial_a / 40;
+        let minimum_token_b_amount = initial_b / 40;
+
+        let mut accounts = SwapAccountInfo::new(
+            &user_key,
+            fees,
+            swap_curve,
+            token_a_amount,
+            token_b_amount,
+            &pool_token_program_id,
+            &token_a_program_id,
+            &token_b_program_id,
+        );
+
+        accounts.initialize_swap().unwrap();
+
+        let (
+            token_a_key,
+            mut token_a_account,
+            token_b_key,
+            mut token_b_account,
+            pool_key,
+            mut pool_account,
+        ) = accounts.setup_token_accounts(
+            &user_key,
+            &withdrawer_key,
+            initial_a,
+            initial_b,
+            initial_pool.try_into().unwrap(),
+        );
+
+        let destination_key = Pubkey::new_unique();
+        let mut destination = SolanaAccount::new(
+            account_minimum_balance(),
+            Account::get_packed_len(),
+            &withdrawer_key,
+        );
+
+        do_process_instruction(
+            close_account(
+                &pool_token_program_id,
+                &accounts.pool_fee_key,
+                &destination_key,
+                &user_key,
+                &[],
+            )
+            .unwrap(),
+            vec![
+                &mut accounts.pool_fee_account,
+                &mut destination,
+                &mut SolanaAccount::default(),
+            ],
+        )
+        .unwrap();
+
+        let user_transfer_authority_key = Pubkey::new_unique();
+        let pool_token_amount = withdraw_amount.try_into().unwrap();
+
+        do_process_instruction(
+            approve(
+                &pool_token_program_id,
+                &pool_key,
+                &user_transfer_authority_key,
+                &withdrawer_key,
+                &[],
+                pool_token_amount,
+            )
+            .unwrap(),
+            vec![
+                &mut pool_account,
+                &mut SolanaAccount::default(),
+                &mut SolanaAccount::default(),
+            ],
+        )
+        .unwrap();
+
+        assert!(do_process_instruction(
```

Comment:
nit: so we can surface the error info more easily on an error, let's do `do_process_instruction(....).unwrap()` rather than asserting that it's ok